src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs

```diff
@@ -0,0 +1,337 @@
+            // (like not ever actually finishing compiling).
+            if (runSubTreeTests)
+            {
+                TryParseSubTrees(stringText, options,
+                    allowIndexOutOfRange,
+                    allowNullReference,
+                    allowOutOfMemory);
+            }
+
+            var actual = TreeToText(tree).Replace("\"", "\"\"");
+            Assert.Equal(expected.Replace("\"", "\"\""), actual);
+        }
+
+        private void TryParseSubTrees(
+            string stringText, RegexOptions options,
+            bool allowIndexOutOfRange ,
+            bool allowNullReference ,
+            bool allowOutOfMemory )
+        {
+            // Trim the input from the right and make sure tree invariants hold
+            var current = stringText;
+            while (current != "@\"\"" && current != "\"\"")
+            {
+                current = current.Substring(0, current.Length - 2) + "\"";
+                TryParseTree(current, options, conversionFailureOk: true,
+                    allowIndexOutOfRange,
+                    allowNullReference,
+                    allowOutOfMemory);
+            }
+
+            // Trim the input from the left and make sure tree invariants hold
+            current = stringText;
+            while (current != "@\"\"" && current != "\"\"")
+            {
+                if (current[0] == '@')
+                {
+                    current = "@\"" + current.Substring(3);
+                }
+                else
+                {
+                    current = "\"" + current.Substring(2);
+                }
+
+                TryParseTree(current, options, conversionFailureOk: true,
+                    allowIndexOutOfRange,
+                    allowNullReference,
+                    allowOutOfMemory);
+            }
+            
+            for (int start = stringText[0] == '@' ? 2 : 1; start < stringText.Length - 1; start++)
+            {
+                TryParseTree(
+                    stringText.Substring(0, start) +
+                    stringText.Substring(start + 1, stringText.Length - (start + 1)),
+                    options, conversionFailureOk: true,
+                    allowIndexOutOfRange,
+                    allowNullReference,
+                    allowOutOfMemory);
+            }
+        }
+
+        private (SyntaxToken, RegexTree, ImmutableArray<VirtualChar>) JustParseTree(
+            string stringText, RegexOptions options, bool conversionFailureOk)
+        {
+            var token = GetStringToken(stringText);
+            var allChars = _service.TryConvertToVirtualChars(token);
+            if (allChars.IsDefault)
+            {
+                Assert.True(conversionFailureOk, "Failed to convert text to token.");
+                return (token, null, allChars);
+            }
+
+            var tree = RegexParser.TryParse(allChars, options);
+            if (tree != null)
+            {
```

Comment:
Missing something?

Reply:
Note: we do not have any ToString support with embedded syntax (haven't needed it yet).  So there is no roundtripping in that sense.   We do validate that all the characters from hte input do show up in the right order in the tree though.

Reply:
Remove.d